modules/xfem/src/efa/EFAFragment2D.C

```diff
@@ -334,345 +334,86 @@ EFAFragment2D::split()
-      new_frag2->addEdge(new EFAEdge(node5, node3));
 
-      new_fragments.push_back(new_frag1);
-      new_fragments.push_back(new_frag2);
-    }
-    else if (cut_edges[0] == 0 && cut_nodes[0] == 3) // case 8
+    if (_boundary_edges[iedge]->numEmbeddedNodes() > 1)
+      EFAError("A fragment boundary edge can't have more than 1 cuts");
+    if (_boundary_edges[iedge]->hasIntersection()) // if current edge is cut add cut intersection //
+                                                   // node to both fragments and and change fragment
     {
-      EFAFragment2D * new_frag1 = new EFAFragment2D(_host_elem, false, NULL);
-      EFAFragment2D * new_frag2 = new EFAFragment2D(_host_elem, false, NULL);
-      EFANode * node1 = _boundary_edges[0]->getNode(0);
-      EFANode * node2 = _boundary_edges[0]->getEmbeddedNode(0);
-      EFANode * node3 = _boundary_edges[0]->getNode(1);
-      EFANode * node4 = _boundary_edges[1]->getNode(1);
-      EFANode * node5 = _boundary_edges[2]->getNode(1);
-
-      new_frag1->addEdge(new EFAEdge(node2, node3));
-      new_frag1->addEdge(new EFAEdge(node3, node4));
-      new_frag1->addEdge(new EFAEdge(node4, node5));
-      new_frag1->addEdge(new EFAEdge(node5, node2));
-
-      new_frag2->addEdge(new EFAEdge(node1, node2));
-      new_frag2->addEdge(new EFAEdge(node2, node5));
-      new_frag2->addEdge(new EFAEdge(node5, node1));
-
-      new_fragments.push_back(new_frag1);
-      new_fragments.push_back(new_frag2);
+      fragment_nodes[frag_number].push_back(_boundary_edges[iedge]->getEmbeddedNode(0));
+      ++edge_cut_count;
+      if (frag_number == 0)
+        frag_number = 1;
+      else
+        frag_number = 0;
+      fragment_nodes[frag_number].push_back(_boundary_edges[iedge]->getEmbeddedNode(0));
     }
   }
-  else if (cut_nodes.size() == 2)
-  {
-    if ((cut_nodes[0] == 0 && cut_nodes[1] == 2) || (cut_nodes[0] == 2 && cut_nodes[1] == 0))
-    {
-      EFAFragment2D * new_frag1 = new EFAFragment2D(_host_elem, false, NULL);
-      EFAFragment2D * new_frag2 = new EFAFragment2D(_host_elem, false, NULL);
-      EFANode * node1 = _boundary_edges[0]->getNode(0);
-      EFANode * node2 = _boundary_edges[1]->getNode(0);
-      EFANode * node3 = _boundary_edges[2]->getNode(0);
-      EFANode * node4 = _boundary_edges[3]->getNode(0);
 
-      new_frag1->addEdge(new EFAEdge(node1, node2));
-      new_frag1->addEdge(new EFAEdge(node2, node3));
-      new_frag1->addEdge(new EFAEdge(node3, node1));
-
-      new_frag2->addEdge(new EFAEdge(node3, node4));
-      new_frag2->addEdge(new EFAEdge(node4, node1));
-      new_frag2->addEdge(new EFAEdge(node1, node3));
-
-      new_fragments.push_back(new_frag1);
-      new_fragments.push_back(new_frag2);
+  if ((edge_cut_count + node_cut_count) > 1) // any two cuts case
+  {
+    EFAFragment2D * new_frag0 = new EFAFragment2D(_host_elem, false, NULL);
+    EFAFragment2D * new_frag1 = new EFAFragment2D(_host_elem, false, NULL);
```

Comment:
I'd love to see smart pointer here. What is the life time management of these objects?

Reply:
I agree. I made an issue to switch everything in this code over to smart pointers about 5 years ago, and still need to get to that...

These are indeed managed correctly (deleted by the elements associated with them), and the code runs cleanly in valgrind, but smart pointers would be nice. That's way beyond the scope of this changeset, though.